Add discrete GPU support for NVIDIA drivers

crawfxrd opened the pull request:
Use data from switcheroo-control to know which environment variables to use to launch an application on the discrete GPU. switcheroo-control version 2.0 or newer should be installed on Linux platforms.

This adds the right click menu "Launch using Dedicated Graphics Card" when using the NVIDIA proprietary drivers.

https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/781

crawfxrd commented:
This had been merged upstream. I'll update the patch will the latest changes.

crawfxrd committed:
Add discrete GPU support for NVIDIA drivers

Use data from switcheroo-control to know which environment variables
to use to launch an application on the discrete GPU. switcheroo-control
version 2.0 or newer should be installed on Linux platforms.

https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/781

crawfxrd commented:
switcheroo-control 2.0 has been packaged in Debian and is available for Ubuntu 20.04 [\[1\]](https://packages.ubuntu.com/focal/switcheroo-control). We will need to backport this to 19.10.

brs17 commented:
> switcheroo-control 2.0 has been packaged in Debian and is available for Ubuntu 20.04 [[1]](https://packages.ubuntu.com/focal/switcheroo-control). We will need to backport this to 19.10.

@crawfxrd , have we backported and included all of the necessary dependencies? Currently when attempting to add this, apt tells me that gnome-shell-common has been held back.

brs17 commented:
This is what I am presented with when running `apt-get install gnome-shell-common`

```
system76@pop-os:~$ sudo apt-get install gnome-shell-common 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  fprintd gir1.2-accountsservice-1.0 gir1.2-gck-1 gir1.2-gcr-3 gir1.2-gdm-1.0 gir1.2-gnomebluetooth-1.0 gir1.2-mutter-5 gir1.2-nm-1.0 gir1.2-nma-1.0 gir1.2-rsvg-2.0 gir1.2-upowerglib-1.0
  gnome-session-bin gnome-session-common gnome-startup-applications gtk2-engines-murrine libfprint0 libpam-fprintd pop-fonts pop-gnome-shell-theme pop-gtk-theme pop-icon-theme
  pop-sound-theme pop-theme switcheroo-control xdg-desktop-portal xwayland
Use 'sudo apt autoremove' to remove them.
The following additional packages will be installed:
  policykit-1-gnome
Recommended packages:
  yaru-theme-gnome-shell
The following packages will be REMOVED:
  chrome-gnome-shell gdm3 gnome-shell gnome-shell-extension-alt-tab-raise-first-window gnome-shell-extension-always-show-workspaces gnome-shell-extension-desktop-icons
  gnome-shell-extension-do-not-disturb gnome-shell-extension-pop-battery-icon-fix gnome-shell-extension-pop-shop-details gnome-shell-extension-pop-suspend-button pop-default-settings
  pop-desktop pop-session
The following NEW packages will be installed:
  policykit-1-gnome
The following packages will be upgraded:
  gnome-shell-common
1 upgraded, 1 newly installed, 13 to remove and 0 not upgraded.
Need to get 731 kB of archives.
After this operation, 8,857 kB disk space will be freed.
Do you want to continue? [Y/n] 
```

apt policy:
```
system76@pop-os:~$ apt policy gnome-shell-common 
gnome-shell-common:
  Installed: 3.34.1-1ubuntu1pop0~1573148724~19.10~f65f9b1
  Candidate: 3.34.3-1ubuntu1~19.10.1pop1~1580917996~19.10~41f9d35
  Version table:
     3.34.3-1ubuntu1~19.10.1pop1~1580917996~19.10~41f9d35 1002
       1002 http://apt.pop-os.org/staging/nvidia-offloading eoan/main amd64 Packages
       1002 http://apt.pop-os.org/staging/nvidia-offloading eoan/main i386 Packages
     3.34.1+git20191024-1ubuntu1~19.10.1 500
        500 http://us.archive.ubuntu.com/ubuntu eoan-updates/main amd64 Packages
        500 http://us.archive.ubuntu.com/ubuntu eoan-updates/main i386 Packages
 *** 3.34.1-1ubuntu1pop0~1573148724~19.10~f65f9b1 1001
       1001 http://ppa.launchpad.net/system76/pop/ubuntu eoan/main amd64 Packages
       1001 http://ppa.launchpad.net/system76/pop/ubuntu eoan/main i386 Packages
        100 /var/lib/dpkg/status
     3.34.1-1ubuntu1 500
        500 http://us.archive.ubuntu.com/ubuntu eoan/main amd64 Packages
        500 http://us.archive.ubuntu.com/ubuntu eoan/main i386 Packages
```


UPDATE: Turns out we needed to add the rest of GNOME 3.34.3 from proposed.

brs17 reviewed (changes requested):
I have had success launching Firefox, Blender, and Steam using the "Launch using dedicated graphics card" selection.

I do notice that the option is in different places for different applications. This can be seen in the following screenshots. Shouldn't this option always be separated from other application selections (such as it is in the following blender screenshot)?


![blender](https://user-images.githubusercontent.com/7460774/73966175-c2fe5000-48d2-11ea-80f7-d82ecbc8f159.png)
![firefox](https://user-images.githubusercontent.com/7460774/73966177-c42f7d00-48d2-11ea-841d-8df3130f2bf7.png)
![steam](https://user-images.githubusercontent.com/7460774/73966178-c42f7d00-48d2-11ea-9919-b6ba08e0a9c5.png)

As expected, this option does not work on all programs. Can we be smart about not showing this option with programs that don't support this?

crawfxrd commented:
> I do notice that the option is in different places for different applications.

This is the current behavior with the nouveau drivers, and the behavior we will see in 20.04 with GNOME 3.36.

"New Window" will always be placed at the top of the context menu (followed by a separator) for unopened apps, unless they override `new-window` in their desktop entry file (e.g., firefox, gedit, evolution).

> Can we be smart about not showing this option with programs that don't support this?

There's no way for GNOME to know if an application supports it. As a follow up, we could use the mechanism introduced in #14 (i.e., applications opt-in to this functionality with an `X-KDE-RunOnDiscreteGpu` setting in their desktop entry file).

brs17 commented:
Note that this does require 3.34.3 that is currently in proposed to also be released.

brs17 commented:
@crawfxrd , right now it appears that this menu option appears not only in Hybrid mode, but also NVIDIA. Since it does not actually do anything in NVIDIA mode, is there a way to make this option not appear?

crawfxrd commented:
gnome-shell could make a dbus call to system76-power to check the mode, and only show the menu if the mode is hybrid.

brs17 reviewed (changes requested):
> gnome-shell could make a dbus call to system76-power to check the mode, and only show the menu if the mode is hybrid.

The option does go away in Intel mode, but I do see it in NVIDIA mode. I originally saw the issue with https://github.com/pop-os/gnome-shell/pull/14 added as well, but have since confirmed this also happens with just this pr.

jackpot51 commented:
There's got to be a reasonable way to know that gnome-shell is running on the NVIDIA card so there is no need to provide additional parameters

crawfxrd commented:
I can probably check the OpenGL vendor and not show the menu item if it's Nouveau or NVIDIA.

crawfxrd committed:
Don't show dGPU launch item if backed by NVIDIA

If GNOME Shell if backed by NVIDIA, don't show the "Launch using
Dedicated Graphics Card" item.

crawfxrd commented:
Will not be added in 19.10. Second commit added to 20.04 in #22.
